### ISSUE TITLE
comptimePrint in comptime block crashes zig test

### ISSUE THREAD

**Issue by silversquirl** [ziglang/zig#14061]
### Zig Version

0.11.0-dev.937+c9e3524d0

### Steps to Reproduce and Observed Behavior

```zig
const std = @import("std");
comptime {
    @compileError(std.fmt.comptimePrint("{}", .{7}));
}
```

```
$ zig test test.zig
Segmentation fault
```

### Expected Behavior

Compile error with message `7`

**Comment by silversquirl** [comment]
Note: this does *not* crash `zig build-exe`; just `zig test`

**Comment by mlugg** [comment]
Further reduction:
```zig
const std = @import("std");

comptime {
    if (std.meta.trait.hasFn("format")(u8)) {}
    @compileError("foo");
}
```
Note that copying `hasFn` into the file and calling *that* works fine, so an uneducated guess is that this may be related to memoization since the test runner is using `fmt`.

**Comment by mlugg** [comment]
Further reduction, this one using `build-exe`. It seems like this might be to do with comptime closure saving parameters past the lifetime of the scope that holds them - in this case, the string `"hi"` is created in `main`,  saved in the closure of `strFn("hi")` which is memoized, and then used again in the `comptime` block after `main` has been fully analyzed.
```zig
fn strFn(comptime name: []const u8) fn () []const u8 {
    const Closure = struct {
        pub fn f() []const u8 {
            return name;
        }
    };
    return Closure.f;
}

pub fn main() void {
    _ = strFn("hi")();
}

comptime {
    @compileError(strFn("hi")());
}
```

**Comment by Vexu** [comment]
Duplicate of #12765, fixed by #13017
